src/transformers/generation/utils.py

```diff
@@ -1497,6 +1482,127 @@ def _supports_default_dynamic_cache(self) -> bool:
         """
         return self._supports_cache_class and "jamba" not in self.__class__.__name__.lower()
 
+    def _prepare_cache_for_generation(
```

Comment:
New function, moving the cache logic from `generate`.

It is doing essentially the same, except for the `Quick escape route 2`, which is new. Added the warning in `Quick escape route 3`.